core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotDownloader.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.raft;
+
+import alluxio.grpc.DownloadSnapshotPRequest;
+import alluxio.grpc.DownloadSnapshotPResponse;
+import alluxio.grpc.SnapshotData;
+import alluxio.grpc.UploadSnapshotPRequest;
+import alluxio.grpc.UploadSnapshotPResponse;
+
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+import org.apache.ratis.io.MD5Hash;
+import org.apache.ratis.server.protocol.TermIndex;
+import org.apache.ratis.server.storage.FileInfo;
+import org.apache.ratis.statemachine.SnapshotInfo;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * A stream observer for downloading a snapshot.
```

Comment:
> Only the leader has the RPC service, right? There are 2 types of transfers:
Yes

> leader receives snapshot from follower: What is the RPC data flow here?  Since the leader only has the RPC service, how does the follower know when to upload to leader? Does the follower just proactively try to transfer to leader whenever it creates a new snapshot?

Here is the workflow:
- ratis calls leader state machine to take a snapshot
- leader get snapshot metadata from follower
- leader pick one of the the follower and send a request for copying the snapshot
- follower receives the request and calls the leader raft journal service to upload the snapshot
- after the upload completes, leader remembers the temporary snapshot location and index.
- ratis calls the leader state machine again to take a snapshot
- leader moves the temporary snapshot to the journal snapshot folder and gives ratis the snapshot index.

> follower receives snapshot from leader: What is the RPC data flow here? When would the follower want to download a snapshot from the leader?

- ratis leader found one of the follower needs a snapshot because it missed journal entries from a long time ago.
- ratis leader notifies ratis follower to install snapshot from leader, the follower calls the alluxio state machine to full fill this request
- the follower state machine calls the snapshot manager which calls the raft journal service from leader to download a snapshot.
- after the downloads completes, follower moves the file to snapshot directory and gives ratis the snapshot index.